Add configurable retry mechanism - #59

SimonJang opened the pull request:
This PR adds a configurable retry mechanism that can be configured in:
- The `.connect()` call when initialising `dynongo`
- During individual operations

The individual configuration takes precedence over the global configured retry options.

This PR includes:

- Changes to all the existing CRUD methods for records to include retry mechanism.
- Add additional tests to validate the above changes.
- Update documentation to demonstrate this new retry functionality.

Note:

The CRUD operations for the tables itself are currently not yet retryable. Issue #59 did not specific mention this feature.

SimonJang commented:
@SamVerschueren 

Just saw that `p-retry` has no support for < Node 8. Want me to downgrade?

SamVerschueren commented:
Maybe we should just ditch older Node.js versions then.

bartcallant commented:
Should there be an option to use the default retry options? Since now it will only support passing in an object.

Also what if an empty object is passed in?

(I'm not up to speed with what you discussed about this already)

SamVerschueren commented on readme.md:
```diff
@@ -50,6 +52,18 @@ db.connect({
 
+#### Retry
+
+Retry configuration can be passed well during the initialisation. The retry mechanism is based on [p-retry](https://github.com/sindresorhus/p-retry) and requires the same [options](https://github.com/tim-kos/node-retry#retryoperationoptions). Configuring the retry will allow the user to automatically retry the dynongo operation if it's an retryable error.
```
```suggestion
The retry configuration can be passed during initialisation, or per individual query. The mechanism is based on [`p-retry`](https://github.com/sindresorhus/p-retry) and requires the same [options](https://github.com/tim-kos/node-retry#retryoperationoptions). Configuring retry will allow the user to automatically retry the DynamoDB operation if it's a retryable error.
```

SamVerschueren commented on readme.md:
````diff
@@ -50,6 +52,18 @@ db.connect({
+
+```js
+db.connect({
+	retryOptions: {
````
Maybe we should just go with `retry` here?

SimonJang replied:
Ok, will change it to a number as well and not the options from `p-retry`

SamVerschueren commented on readme.md:
````diff
@@ -94,6 +108,15 @@ const Employee = db.rawTable('Employee');
+Every method can override the retry [options](https://github.com/tim-kos/node-retry#retryoperationoptions) passed with the `.connect()` method or can customise the retry configuration for the specific method.
+
+```js
+Employee.find({Organisation: 'Amazon'}).where({Salary: {$gt: 3000}}).select('FirstName Name').retry({retries: 2}).exec()
````
```suggestion
Employee
	.find({Organisation: 'Amazon'})
	.where({Salary: {$gt: 3000}})
	.select('FirstName Name')
	.retry({retries: 2})
	.exec()
```

SamVerschueren commented on src/lib/methods/base-query.ts:
```diff
@@ -122,6 +124,17 @@ export abstract class BaseQuery extends Method {
+	 *
+	 * @param opts - Retry configuration options.
+	 */
+	retry(opts: Options) {
```
I guess it would make sense to have support only the number of attempts.

```ts
retry(retries: number);
retry(options: Options);
```

I guess, most of the time the only thing you want to do is change the number of retries.

Maybe even 

```ts
retry(retries: number, options?: Options)
```

Not sure about this one. Feedback is welcome.

SimonJang replied:
Will go for number only, seems like the most user-friendly approach here. I thought that providing the full options for the retry strategy might be something that most users would want but in hindsight, it is more something for power users I guess.

SamVerschueren replied:
I think we should support both though... Using overrides makes things easy. In some use cases you might want to adjust the timings etc. So we should definitely expose the full options object.

SamVerschueren commented on src/lib/methods/delete-item.ts:
```diff
@@ -1,13 +1,16 @@
 import { DeleteItemInput } from 'aws-sdk/clients/dynamodb';
+import retry, { Options } from 'p-retry';
```
Rename `Options` to `RetryOptions` here.

SimonJang replied:
Will remove `p-retry` options as argument. Only accepting numbers now to make the API more user friendly.

SamVerschueren commented on src/lib/methods/delete-item.ts:
```diff
@@ -81,7 +95,16 @@ export class DeleteItem extends Method implements Executable {
+		 * Fallback to the retry options declared when db was initialized if defined
+		 * If none of the above is defined, no retry mechanic is used
+		 */
+		const queryResult = this.retryOptions || this.dynamodb.retryOptions
```
Should we merge these 🤔? Not sure.

SimonJang replied:
Fixed in the method `runQuery`

SamVerschueren commented on src/lib/methods/delete-item.ts:
```diff
@@ -81,7 +95,16 @@ export class DeleteItem extends Method implements Executable {
+		 * If none of the above is defined, no retry mechanic is used
+		 */
+		const queryResult = this.retryOptions || this.dynamodb.retryOptions
+			? retry(() => db.delete(this.buildRawQuery()).promise().catch(retryErrorHandler), this.retryOptions)
```
Can we use `async/await` here instead?

I'm wondering if we can make all of this a util function?

```ts
const queryResult = runQuery(() => {
	return db.delete(this.buildRawQuery()).promise();
}, this.retryOptions || this.dynamodb.retryOptions);
```

The implementation picks the execution based on the input of the `retryOptions`.

SimonJang replied:
Made a method in the `Method` class that can execute with or without retry.

SamVerschueren commented on src/lib/methods/delete-item.ts:
```diff
@@ -81,7 +95,16 @@ export class DeleteItem extends Method implements Executable {
+		 * If none of the above is defined, no retry mechanic is used
+		 */
+		const queryResult = this.retryOptions || this.dynamodb.retryOptions
+			? retry(() => db.delete(this.buildRawQuery()).promise().catch(retryErrorHandler), this.retryOptions)
```
I also noticed that you don't use the `this.dynamodb.retryOptions` fallback here.

SimonJang replied:
Fixed

SamVerschueren commented on src/lib/methods/insert-item.ts:
```diff
@@ -72,6 +75,17 @@ export class InsertItem extends Method implements Executable {
+	 *
+	 * @param opts - Retry configuration options.
+	 */
+	retry(opts: Options) {
```
Is this something we can move to the `Method` class? As it feels like it's part of every database method?

SimonJang replied:
Done and the `runQuery` utility function as well.

bartcallant commented:
When @SimonJang and I were talking out retrying the calls to DynamoDB, he showed me [the list errors Amazon provides](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html). And I was wondering if we should extends the errors which are checked in order to determine if the action should be retried. Since this PR already checks more errors than we do internally. The list states which errors are `OK to retry?` according to Amazon.

#### OK to retry? Yes
- ItemCollectionSizeLimitExceededException ✅
- LimitExceededException ✅
- ProvisionedThroughputExceededException ✅
- RequestLimitExceeded ✅
- ThrottlingException ✅
- UnrecognizedClientException ❌
- Internal Server Error (HTTP 500) ✅
- Service Unavailable (HTTP 503) ✅

#### OK to retry? No
- AccessDeniedException ❌
- ConditionalCheckFailedException ❌
- IncompleteSignatureException ❌
- MissingAuthenticationTokenException ❌
- ResourceInUseException ✅
- ResourceNotFoundException ❌
- ValidationException ❌

_✅ indicates which I think should be retried, ❌ obviously no? 😛_

This is not a request to change it, but more like a question/suggestion.

SamVerschueren commented:
@bartcallant 

**UnrecognizedClientException**
Not sure about this one. But might be something that could be resolved by retrying? Hard to say when it's throwed though.

All the other look good! Thanks for looking into this!

SimonJang commented:
PR updated!

SamVerschueren commented on readme.md:
```diff
@@ -94,6 +118,34 @@ const Employee = db.rawTable('Employee');
+	.find({Organisation: 'Amazon'})
+	.where({Salary: {$gt: 3000}})
+	.select('FirstName Name')
+	.retry({retries: 2})
```
Provide the same options as above to make it clear that it accepts all the p-retry stuff.

```suggestion
	.retry({retries: 3, factor: 1, randomize: false})
```

SamVerschueren commented on readme.md:
```diff
@@ -94,6 +118,34 @@ const Employee = db.rawTable('Employee');
+	.retry({retries: 2})
+	.exec()
+    .then(employees => {
+        // => [{FirstName: 'Foo', Name: 'Bar'}]
```
Indentation

SimonJang replied:
Could you give an suggestion on what the identation needs to be? It has the same identation as other documentation (see line 154)

SamVerschueren replied:
Because of refined GitHub I noticed that it has spaces instead of tabs. Couldn't see the difference on the mobile app.

![image](https://user-images.githubusercontent.com/1913805/73597165-84127800-4529-11ea-8f44-4b61f746a7e5.png)

SamVerschueren commented on readme.md:
```diff
@@ -94,6 +118,34 @@ const Employee = db.rawTable('Employee');
+	.exec()
+    .then(employees => {
+        // => [{FirstName: 'Foo', Name: 'Bar'}]
+    });
```
Indentation

SimonJang replied:
Could you give an suggestion on what the identation needs to be? It has the same identation as other documentation (see line 154)

SamVerschueren replied:
It looked off when I reviewed it. On mobile now and it looks ok.

SamVerschueren replied:
Same, it has spaces instead of tabs.

SamVerschueren commented on src/lib/methods/method.ts:
```diff
@@ -16,4 +19,23 @@ export abstract class Method implements QueryBuilder {
+	/**
+	 * Configure the retry policy
+	 *
+	 * @param opts - Retry configuration options.
```
```suggestion
	 * @param retries - Retry configuration options.
```

SamVerschueren replied:
If you want, you can actually work with overrides.

```ts
retry(retries: number);
retry(options: RetryOptions);

retry(options: number | RetryOptions) {
	this.retries = options;

	return this;
}
```

Only the first two will end up in intellisense so you can document them differently. It's just a TS feature, so feel free to use it. But if you want, you may keep the current implementation.

SimonJang replied:
When trying to implement your suggestion, the linter is complaining. Can I add an exception? 

`ERROR: 34:8  unified-signatures      These overloads can be combined into one signature taking `number | RetryOptions`.`

SamVerschueren replied:
Yup, just disable it :).

SamVerschueren commented on src/lib/dynamodb.ts:
```diff
@@ -14,13 +15,15 @@ export interface DynamoDBOptions {
 	public raw?: AWS.DynamoDB;
 	public dynamodb?: AWS.DynamoDB.DocumentClient;
 	private options: DynamoDBOptions = {};
+	public retries?: number | RetryOptions;
```
Maybe we just make this private and use a `get retries()`? That way, nobody else can change the value.

SamVerschueren commented on src/lib/methods/method.ts:
```diff
@@ -16,4 +19,23 @@ export abstract class Method implements QueryBuilder {
+		const retries = this.retries || this.dynamodb.retries;
+
+		return retries
+			? retry(() => operation().catch(retryErrorHandler), {...typeof retries === 'number' ? {retries} : retries})
```
Extracting the retry options to a variable will be easier to read I guess.

```ts
const retryOptions = typeof retries === 'number' ? {retries} : retries;
```

SamVerschueren replied:
I wasn't sure about it, but did we discuss passing default options to `p-retry`? Like the randomizer thing etc? I think we should come up with some sensible defaults.

SimonJang replied:
The current implementation uses the defaults from the `p-retry` library and the `retry` mechanism is an opt-in feature, it's not enabled by default.

SamVerschueren replied:
I know, but if we do `retry(5)`, the exponential back off thing should make sense and we should add some random overwriting the defaults from `retry`. Let's discuss tomorrow f2f :).

SamVerschueren commented on src/lib/utils/retry-error-handler.ts:
```diff
@@ -0,0 +1,20 @@
+import { AbortError } from 'p-retry';
+
+const whitelistedErrors = {
```
Use a `Set` for this so you can use `.has(err.code)`.

SimonJang commented:
PR updated

SimonJang commented:
PR updated. We only need to decide on retry defaults.

SamVerschueren commented on src/lib/utils/retry-options.ts:
```diff
@@ -0,0 +1,22 @@
+		return;
+	}
+
+	return typeof retries === 'number' ? {...defaultRetryOptions, retries} : {...defaultRetryOptions, ...retries};
```
Can you spread the ternary over multiple lines?

```ts
return typeof retries === 'number'
	? {...defaultRetryOptions, retries}
	: {...defaultRetryOptions, ...retries};
```

SamVerschueren commented on src/lib/utils/retry-options.ts:
```diff
@@ -0,0 +1,22 @@
+
+const defaultRetryOptions: RetryOptions = {
+	retries: 5,
+	factor: 2,
```
I think we decided to go with `1`, so no exponential backoff.

SamVerschueren commented on src/lib/utils/retry-options.ts:
```diff
@@ -0,0 +1,22 @@
+const defaultRetryOptions: RetryOptions = {
+	retries: 5,
+	factor: 2,
+	minTimeout: 500,
```
If we go with `300`, the possible emits will happen at the following times
- [300, 600]
- [600, 1200]
- [900, 1800]
- [1200, 2000]
- [1500, 2000]

If we go with `500`
- [500, 1000]
- [1000, 2000]
- [1500, 2000]
- [2000, 2000]
- [2000, 2000]

When reading the `retry` docs, it looks like the emits are calculated individually. This means that in case of `300`, it's possible to the first retry happens after `600ms` (if random is `2`), and the second retry happens directly after that one (if random is `1`). Not sure about this behaviour though.

So looking at the times, I believe we should go with `300` here. What do you think?

SimonJang commented:
Updated the PR. I changed the readme files to have tabs in the examples since most of the examples used spaces and you made a comment about my indentation.

SamVerschueren commented on readme.md:
```diff
@@ -94,40 +118,68 @@ const Employee = db.rawTable('Employee');
+	.retry(2)
+	.exec()
+	.then(employees => {
+	// => [{FirstName: 'Foo', Name: 'Bar'}]
```
```suggestion
		// => [{FirstName: 'Foo', Name: 'Bar'}]
```

SamVerschueren commented:
Jeej 🎉! I'm so glad we landed this!

![cheers](https://user-images.githubusercontent.com/1913805/74807982-19638980-52ea-11ea-9634-945b1da88bf8.gif)
